src/event/merge.rs

```diff
@@ -25,10 +26,26 @@ pub fn merge_value(current: &mut Value, incoming: Value) {
     }
 }
 
+/// Merges `source` map into `dest` recursively
+pub fn merge_nested(dest: &mut HashMap<Atom, Value>, source: HashMap<Atom, Value>) {
```

Comment:
"merge" has different semantics compared to other functions in this mod - in a sense that it doesn't merge the inner field values. Consider adding more explanatory documentation to avoid confusing the users.

Also, `merge_log_event` should probably rely on the logic provided by this function in some form, but the support for merging messages in nested events is not implemented altogether. Maybe add some closure to specify the action to take on the field.
This has to be addressed if we want to support merging values in nested fields - and not supporting it would be a regression.

Lastly, my gut feeling tells me it should be possible to avoid special-casing if we make an `Event` to hold a `fields: Value,` instead of `fields: HashMap<Atom, Value>,` - then the special casing here and at `Event#insert` will go away. What do you think?

Reply:
I got the point about the naming, I'll try to change it to make it less confusing.

> Lastly, my gut feeling tells me it should be possible to avoid special-casing if we make `LogEvent` hold a `fields: Value,` instead of `fields: HashMap<Atom, Value>,` - then the special casing here and at `Event#insert` will go away. What do you think?

I was thinking about this too. It might look more elegant, but if we made `fields` a `Value`, then it would be necessary to always check if `fields` is a map, which would decrease performance and require us to specify behavior for log events for which `fields` is a scalar or an array.

Reply:
Right, that would complicate things unnecessarily...
Btw, have you considered adding a type alias for `HashMap<Atom, Value>`? It seems like a good idea - for example, if we'd need to switch to `BTreeMap`, we'll just update the alias instead of changing the types all over the code.

Reply:
>  Btw, have you considered adding a type alias for `HashMap<Atom, Value>`? It seems like a good idea - for example, if we'd need to switch to `BTreeMap`, we'll just update the alias instead of changing the types all over the code.

This sounds reasonable to me.

Reply:
Added #1807 to ensure we address the issue with the merging of the nested fields.